plotters-svg/src/svg.rs

```diff
@@ -80,36 +77,162 @@ pub struct SVGBackend<'a> {
+        };
     }
-    fn open_tag(&mut self, tag: SVGTag, attr: &[(&str, &str)], close: bool) {
-        let buf = self.target.get_mut();
+}
+
+struct FormatEscapedIter<I>(I);
+impl<I: IntoIterator<Item: FormatEscaped>> FormatEscaped for FormatEscapedIter<I> {
+    fn format_escaped(buf: &mut String, iter: FormatEscapedIter<I>) {
+        let iter = iter.0.into_iter();
+        for item in iter {
+            FormatEscaped::format_escaped(buf, item);
+        }
+    }
+}
+
+enum Value {}
+enum Init {}
+struct AttrWriter<'a, State> {
+    buf: &'a mut String,
+    tag: SVGTag,
+    tag_stack: &'a mut Vec<SVGTag>,
+    state: std::marker::PhantomData<State>,
+}
+
+/// Used for opening a tag and then optionally writing some attributes. The expected workflow is
+/// to call [open_tag](AttrWriter::open_tag), then zero or more times calling
+/// [write_key](AttrWriter::write_key) followed optionally by `[write_value](AttrWriter::write_value)`
+/// and finally calling one of [close](AttrWriter::close) (to close a self-closing tag)
+/// or [finish_without_closing](AttrWriter::finish_without_closing) (to schedule writing
+//  writing the closing tag for later).
+impl<'a> AttrWriter<'a, Init> {
+    fn open_tag(buf: &'a mut String, tag: SVGTag, tag_stack: &'a mut Vec<SVGTag>) -> Self {
         buf.push('<');
         buf.push_str(tag.to_tag_name());
-        for (key, value) in attr {
-            buf.push(' ');
-            buf.push_str(key);
-            buf.push_str("=\"");
-            Self::escape_and_push(buf, value);
-            buf.push('\"');
-        }
-        if close {
-            buf.push_str("/>\n");
-        } else {
-            self.tag_stack.push(tag);
-            buf.push_str(">\n");
+        AttrWriter {
+            buf,
+            tag,
+            tag_stack,
+            state: Default::default(),
         }
     }
 
+    fn write_key<'s>(&'s mut self, key: &str) -> AttrWriter<'s, Value> {
+        self.buf.push(' ');
+        self.buf.push_str(key);
+        AttrWriter {
+            buf: self.buf,
+            tag: self.tag.clone(),
+            tag_stack: &mut self.tag_stack,
+            state: Default::default(),
+        }
+    }
+
+    fn close(self) {
+        self.buf.push_str("/>\n");
+    }
+
+    fn finish_without_closing(self) {
+        self.tag_stack.push(self.tag);
+        self.buf.push_str(">\n");
+    }
+}
+
+impl<'a> AttrWriter<'a, Value> {
+    fn write_value(self, value: impl FormatEscaped) {
+        self.buf.push_str("=\"");
+        FormatEscaped::format_escaped(self.buf, value);
+        self.buf.push('"');
+    }
+}
+
+impl<'a> SVGBackend<'a> {
+    fn escape_and_push(buf: &mut String, value: &str) {
+        value.chars().for_each(|c| match c {
+            '<' => buf.push_str("&lt;"),
+            '>' => buf.push_str("&gt;"),
+            '&' => buf.push_str("&amp;"),
+            '"' => buf.push_str("&quot;"),
+            '\'' => buf.push_str("&apos;"),
+            other => buf.push(other),
```

Comment:
This is duplicated code, you could reuse the implementation of `FormatEscaped` for `char`